src/datasets/builder.py

```diff
@@ -1467,43 +1538,125 @@ def _prepare_split(
+            num_proc = 16
+
+        if _shard_number(split_generator.gen_kwargs) <= 1 and num_proc is not None:
+            logger.warning(
+                f"Setting num_proc from {num_proc} back to 1 to disable multiprocessing as dataset is not shardable"
+            )
+            num_proc = None
+
+        if num_proc is None or num_proc == 1:
+            result = self._prepare_split_single(
+                split_generator.gen_kwargs,
+                fpath=fpath,
+                file_format=file_format,
+                max_shard_size=max_shard_size,
+                split_info=split_info,
+            )
+            # wrapping everything into lists for consistency with the multiprocessed code path
+            examples_per_rank, bytes_per_rank, features_per_rank, shards_per_rank = [[item] for item in result]
+        else:
+            args_per_shard = [
+                (
+                    shard_kwargs,
+                    fpath,
+                    file_format,
+                    max_shard_size,
+                    split_info,
+                )
+                for shard_kwargs in _all_shard_kwargs(split_generator.gen_kwargs)
+            ]
+
+            examples_per_rank = [None] * len(args_per_shard)
+            bytes_per_rank = [None] * len(args_per_shard)
+            features_per_rank = [None] * len(args_per_shard)
+            shards_per_rank = [None] * len(args_per_shard)
+
+            with Pool(num_proc) as pool:
+                results = {
+                    rank: pool.apply_async(self._prepare_split_single, args=args, kwds={"rank": rank})
+                    for rank, args in enumerate(args_per_shard)
+                }
+                for index, async_result in results.items():
+                    result = async_result.get()
+                    (
+                        examples_per_rank[index],
+                        bytes_per_rank[index],
+                        features_per_rank[index],
+                        shards_per_rank[index],
+                    ) = result
+
+            assert None not in examples_per_rank, f"result list {examples_per_rank} still contains None"
+            # wrapping everything into lists for consistency with the multiprocessed code path
+
+        total_shards = sum(shards_per_rank)
+        total_num_examples = sum(examples_per_rank)
+        total_num_bytes = sum(bytes_per_rank)
+        features = features_per_rank[0]
+
+        # should rename everything at the end, scheme still TBD
+        def _rename_shard(shard_id_and_rank: Tuple[int]):
+            shard_id, rank = shard_id_and_rank
+            global_shard_id = sum(shards_per_rank[:rank]) + shard_id
+            self._rename(
+                fpath.replace("SSSSS", f"{shard_id:05d}").replace("RRRRR", f"{rank:05d}"),
+                fpath.replace("RRRRR-SSSSS", f"{global_shard_id:05d}").replace("NNNNN", f"{total_shards:05d}"),
+            )
```

Comment:
Do you think we can preserve the order of the dataset with a smart way to rename those shards ?

If every arrow shard contains the index of the args_per_shard that was used, I think we can get the order of the original dataset back.

This is not a super high priority though

Reply:
It does! Or at least, this preserves the order of the shards in `split_generator.gen_kwargs`.

Reply:
Actually it doesn't after testing, but I can't quite figure out why :/